### PR TITLE
WT-9814 Generate dSYM bundles for macOS core dumps

### PR DESCRIPTION
**Pull request by Sean04**
I've added cmake custom commands to the target test executables and wiredtiger shared library in order to generate dSYM bundles using the dsymutil on macOS. This provides a way to load the symbols previously missing in macOS core dumps from evergreen.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -210,6 +210,17 @@ if(ENABLE_SHARED)
     )
     # Set the SOVERSION property of the wiredtiger library, so we can export the appropriately versioned symlinks.
     set_target_properties(wiredtiger_shared PROPERTIES SOVERSION "${VERSION_MAJOR}.${VERSION_MINOR}.${VERSION_PATCH}")
+
+    # For MacOS builds we need to generate a dSYM bundle that contains the debug symbols for each
```
Comment (korteland) [inline CMakeLists.txt:214]: Nitpick, I think the "each" got left over from the other comment.

File: CMakeLists.txt [hunk 2]
```diff
@@ -210,6 +210,17 @@ if(ENABLE_SHARED)
     )
     # Set the SOVERSION property of the wiredtiger library, so we can export the appropriately versioned symlinks.
     set_target_properties(wiredtiger_shared PROPERTIES SOVERSION "${VERSION_MAJOR}.${VERSION_MINOR}.${VERSION_PATCH}")
+
+    # For MacOS builds we need to generate a dSYM bundle that contains the debug symbols for each 
+    # the WiredTiger library.
+    if (WT_DARWIN AND HAVE_DIAGNOSTIC)
```
Comment (korteland) [inline CMakeLists.txt:216]: I think just `HAVE_DARWIN` would be better here. We could (in theory) have debug symbols without diagnostics.

File: CMakeLists.txt [hunk 3]
```diff
@@ -210,6 +210,17 @@ if(ENABLE_SHARED)
...
+        add_custom_command(
+            TARGET wiredtiger_shared POST_BUILD
+            COMMAND dsymutil libwiredtiger.${VERSION_MAJOR}.${VERSION_MINOR}.${VERSION_PATCH}.dylib
```
Comment (korteland) [inline CMakeLists.txt:219]: How does this end up working in practice? Do I need to load two dSYMs, one for the library and one for the binary?

File: test/ctest_helpers.cmake [hunk 4]
```diff
@@ -37,6 +37,29 @@ function(create_test_executable target)
...
+                COMMAND dsymutil ${target}
+                WORKING_DIRECTORY ${test_binary_dir}
+                COMMENT "Running dsymutil on ${target}"
+                VERBATIM
+            )
+        else()
+            add_custom_command(
+                TARGET ${target} POST_BUILD
+                COMMAND dsymutil ${CREATE_TEST_EXECUTABLE_NAME}
+                WORKING_DIRECTORY ${test_binary_dir}
```
Comment (korteland) [inline test/ctest_helpers.cmake:56]: I understood this when you explained it to me, but reading the code - there's a lot of duplication here, is there a downside to using `${target}` for both?

**Review by korteland — COMMENTED** [review]
Overall idea/approach is sound - thanks for fixing this. Got a couple of comments and questions.

File: CMakeLists.txt (hunk 2 above)
Comment (Sean04) [reply CMakeLists.txt:216]: Yeah fair enough, but do you mean just `WT_DARWIN` or is `HAVE_DARWIN` a thing?

Comment (korteland) [reply CMakeLists.txt:216]: Yeah I sneakily fixed it. You caught me anyway, though.

File: CMakeLists.txt (hunk 3 above)
Comment (Sean04) [reply CMakeLists.txt:219]: In practice LLDB will look for the corresponding dSYM files on the machine. So after my first test of generating the dSYM bundle for the shared library the subsequent tests would automatically find that same bundle. So this is still necessary for having the correct symbols for that commit of wiredtiger.

File: test/ctest_helpers.cmake (hunk 4 above)
Comment (Sean04) [reply test/ctest_helpers.cmake:56]: I'll make a temporary variable and set it to either `target` or `CREATE_TEST_EXECUTABLE_NAME` then use that temporary variable in the `add_custom_command`

Comment (korteland) [reply test/ctest_helpers.cmake:56]: That works too, but my question is more about whether we can just use `${target}` for both.

File: CMakeLists.txt (hunk 3 above)
Comment (korteland) [reply CMakeLists.txt:219]: Okay. It sounds like we might need to manually load two bundles when debugging with downloaded artifacts then?

(That's not a problem and we already have to do something similar with GDB's `set solib-search-path`, I just want to confirm my understanding.)

Comment (Sean04) [reply CMakeLists.txt:219]: Yeah, I believe so.

File: test/ctest_helpers.cmake (hunk 4 above)
Comment (Sean04) [reply test/ctest_helpers.cmake:56]: The problem is we need to use the name of the executable file that's eventually created. So just below the `add_custom_command` we set the `OUTPUT_NAME` property of the target executable which renames the executable file in the end. If it's possible to access that property as a variable in the `add_custom_command` we could move the command after we set that property and just the OUTPUT_NAME as the variable. I'm not sure if that's possible though.

**Review by korteland — APPROVED** [review]
LGTM - nice work on this!

File: CMakeLists.txt [hunk 5]
```diff
@@ -210,6 +210,17 @@ if(ENABLE_SHARED)
     )
     # Set the SOVERSION property of the wiredtiger library, so we can export the appropriately versioned symlinks.
     set_target_properties(wiredtiger_shared PROPERTIES SOVERSION "${VERSION_MAJOR}.${VERSION_MINOR}.${VERSION_PATCH}")
+
+    # For MacOS builds we need to generate a dSYM bundle that contains the debug symbols for the 
+    # WiredTiger library.
+    if (WT_DARWIN)
```
Comment (lukech) [inline CMakeLists.txt:216]: Do we need to check the enabling of the diagnostics flag as well?

Comment (Sean04) [reply CMakeLists.txt:216]: I actually had that check before but @korteland suggested having the debug symbols without diagnostics in the above comments.

File: test/ctest_helpers.cmake [hunk 6]
```diff
@@ -37,6 +37,26 @@ function(create_test_executable target)
 
     # Define our test executable.
     add_executable(${target} ${CREATE_TEST_SOURCES})
+
+    # For MacOS builds we need to generate a dSYM bundle that contains the debug symbols for each 
+    # executable. The name of the binary will either be the name of the target or some other name
+    # passed to this function. We need to use the correct one for the dsymutil.
```
Comment (lukech) [inline test/ctest_helpers.cmake:43]: Do we have a rough idea of how large the sum of the bundle files could be for all test executables? Just hope to make sure it's not going to cause disk space issues on the test host.

Comment (Sean04) [reply test/ctest_helpers.cmake:43]: Good question; the .dSYM bundle for the shared library and test format are about ~600 Bytes each so each bundle is small.

Comment (Sean04) [reply test/ctest_helpers.cmake:43]: Good question; the bundle size for the shared library is ~11MB and the bundle for test format (as an example) is 718KB so I think that's small enough.

File: CMakeLists.txt (hunk 5 above)
Comment (lukech) [reply CMakeLists.txt:216]: Got it, it's okay with me. I was just confused after reading the wiki page (the referring enable -HAVE_DIAGNOSTIC is a requirement).

File: test/ctest_helpers.cmake (hunk 6 above)
Comment (lukech) [reply test/ctest_helpers.cmake:43]: Good to know - that level of size is perfectly fine.

**Review by lukech — APPROVED** [review]
LGTM

File: CMakeLists.txt (hunk 5 above)
Comment (Sean04) [reply CMakeLists.txt:216]: Ah good point, I'll remove it there too then.

**Comment by Sean04** [comment]
evergreen merge

**Comment by Sean04** [comment]
evergreen merge

**Pull request merged**